File: ctags/perl/tags.pl
#!/usr/bin/perl

use strict;
use File::Basename;
use Getopt::Long;

my $notsorted;
my $langOpt = '';

GetOptions ("notsorted" => \$notsorted,
            "language=s" => \$langOpt,
            )
  or Usage();

my $basedir = dirname($0);
$basedir = "." if ($basedir eq "");

my $filename = shift;
my $output = shift;


Usage() if $filename eq "";

#print STDERR "Tokenizing $filename\n";
if ($output ne "") {
    open(OUT, ">$output") or die "Unable to create output file\n";
    select OUT;
}

Declarations($filename, $notsorted);

if ($output ne "") {
    close(OUT);
}

exit;

sub Declarations
{
    my ($filename, $notsorted) = @_;

    #    my $CTAGS = "ctags-exuberant --language-force=c -x";

    my $CTAGS = 'ctags -x ';
    if ($langOpt ne "") {
        $CTAGS .= "--language-force='$langOpt' ";
    }

    if ($notsorted) {
        $CTAGS = $CTAGS . " -u";
    }

    my @declarations;

    open(ctags, "$CTAGS '$filename'|") or die "Unable to execute ctags on file [$filename]";

    while (<ctags>) {
        my %decl;
        my $rest;
        my $line;
        chomp;
        $decl{original} = $_;
        die "unable to parse output ctags [$_]" unless /^(.+?)\s+([^ ]+?)\s+([0-9]+)\s+(.+)$/;
        ($decl{name}, $decl{type}, $decl{line}, $rest) = ($1, $2, $3, $4);
        die "illegal type [$decl{type}] [$_] $decl{name}" unless $decl{type} eq "function" or
                $decl{type} eq "class" or
                $decl{type} eq "macro" or
                $decl{type} eq "member" or
                $decl{type} eq "method" or
                $decl{type} eq "field" or
                $decl{type} eq "constant" or
                $decl{type} eq "package" or
                $decl{type} eq "label" or
                $decl{type} eq "namespace" or
                $decl{type} eq "define" or
                $decl{type} eq "type" or
                $decl{type} eq "enumerator" or
                $decl{type} eq "subroutine" or
                $decl{type} eq "struct" or
                $decl{type} eq "typedef" or
                $decl{type} eq "enum" or
                $decl{type} eq "union" or
                $decl{type} eq "variable" ;

        push @declarations, "DECL|$decl{type}|$decl{name}\n";
    }
    if ($notsorted) {
        print @declarations;
    } else {
        print sort @declarations;
    }
    close ctags;

}

sub Usage {
    print STDERR @_ if @_;
    die "$0 [options]<filename> <outputfile*>\n\n\toptions: --sorted";
}

File: ctags/python/tags.py
#!/usr/bin/python3

# author Aria Poutanen
#

import sys
import subprocess
import linecache

types = ["prototype", "function", "class", "macro", "member", "method", "field", "constant", "package", "label", "namespace", "define", "type", "enumerator", "subroutine", "struct", "typedef", "enum", "union", "variable"]

if len(sys.argv) < 2 or len(sys.argv) > 3:
    print("tags.py <filename> <outputfile*>")
    exit()

filename = sys.argv[1]
if (filename == '--language=C'):
   filename = sys.argv[2]

sys.stderr.write(f"Tokenizing {filename}\n")

if len(sys.argv) >= 4:
    sys.stdout = open(sys.argv[2], 'w')

output = subprocess.run(['ctags', "--language-force=C", '-x', '--c-kinds=+p', filename], stdout=subprocess.PIPE).stdout.decode('utf-8').split('\n')
arr = []

for line in output:

    line = line.split()

    if len(line) < 5:
        # end of ctags output
        break

    name = line[0]
    token_type = line[1]
    line_num = line[2]
    file_name = line[3]
    decl = " ".join(line[4:])

    # check types
    if token_type not in types:
        sys.stderr.write(f"illegal type {token_type} in {name}")
        exit (1)

    # each line is now of the formal [function name, type, line number, file, declaration]
    # inspect declaration to see if it is truncated or not

    stack = []
    eol = False;
    parens = False;
    for char in decl:
        if char == '(':
            parens = True;
            stack.append(char)
        elif char == ')':
            if len(stack) > 0:
                stack.pop()
            else:
                print("Unmatched braces in function declaration")
                exit(1)
        elif char == ';':
            eol = True;

    if ((token_type == 'function') or (token_type == 'prototype')) and ((len(stack) > 0) or (not parens and not eol)):
        # need to search in file for full function decl
        lines = []
        with open(file_name, encoding='ISO-8859-1') as f:
            lines = f.readlines()
        loop = True
        i = 0
        while(loop):
            if i > 100000:
                # we don't want infinite recursion
                break
            next_line = lines[int(line_num)+i]
            for char in next_line:
                if (char == ';') or (char == '{'):
                    loop = False
                    break
                else:
                    decl = decl + char
            i = i+1

        # replace tabs and whitespace
        decl = decl.replace('\t','')
        decl = " ".join(decl.split())

    print_decl = "";

    if token_type == "variable" or token_type == "macro":
        # normally we would consider macros, but we are only looking at functions or prototypes for this case
        # note to update this for script reuse
        print_decl = name;
    else:
       #print_decl = decl[decl.find(name):]
        print_decl = decl

    arr.append(f"DECL|{name}|{token_type}|{print_decl}")

print('\n'.join(sorted(arr)))
